aimmo-game/simulation/action.py

```diff
@@ -1,145 +1,152 @@
-            return self.apply(world_map)
+        avatar_id = game_state.avatar_at(self.target)
 
-        next_action = next_cell.avatar.action
-        if next_action.chain(world_map, visited | {self.target_location}):
-            return self.apply(world_map)
+        return (other_actions is not None
+                and other_actions.avatar_moving(avatar_id))
 
-        return self.reject()
+    def apply(self, game_state, other_actions):
+        game_state.move_avatar(self.avatar_id, self.direction)
+        game_state.add_event(self.avatar_id, MovedEvent(self.origin, self.target))
+        return True
 
-    def reject(self):
-        event = FailedMoveEvent(self.avatar.location, self.target_location)
-        self.avatar.add_event(event)
-        self.avatar.clear_action()
+    def reject(self, game_state):
+        game_state.add_event(self.avatar_id, FailedMoveEvent(self.origin, self.target))
         return False
 
 
 class AttackAction(Action):
-    def __init__(self, avatar, direction):
-        # Untrusted data!
-        self.direction = Direction(**direction)
-        super(AttackAction, self).__init__(avatar)
-
-    def is_legal(self, world_map):
-        return True if world_map.attackable_avatar(self.target_location) else False
-
-    def apply(self, world_map):
-        attacked_avatar = world_map.attackable_avatar(self.target_location)
-        damage_dealt = 1
-        self.avatar.add_event(PerformedAttackEvent(attacked_avatar,
-                                                   self.target_location,
-                                                   damage_dealt))
-        attacked_avatar.add_event(ReceivedAttackEvent(self.avatar,
-                                                      damage_dealt))
-        attacked_avatar.health -= damage_dealt
-
-        LOGGER.debug('{} dealt {} damage to {}'.format(self.avatar,
-                                                       damage_dealt,
-                                                       attacked_avatar))
-        self.avatar.clear_action()
-
-        if attacked_avatar.health <= 0:
-            # Move responsibility for this to avatar.die() ?
-            respawn_location = world_map.get_random_spawn_location()
-            attacked_avatar.die(respawn_location)
-            world_map.get_cell(self.target_location).avatar = None
-            world_map.get_cell(respawn_location).avatar = attacked_avatar
-
-    def reject(self):
-        self.avatar.add_event(FailedAttackEvent(self.target_location))
-        self.avatar.clear_action()
+    def __init__(self, avatar, source, direction):
+        #                                               Untrusted data!
+        Action.__init__(self, avatar, source, direction=Direction.copy(direction))
+
+    def _attacked_avatar(self, game_state, other_actions):
+        if game_state.cell_occupied(self.target):
+            return game_state.avatar_at(self.target)
+        elif (other_actions is not None
```

Comment:
Can we pass in an empty ActionRegistry rather than having to check for None everywhere? This has a high probability of going wrong if someone later modifies the code and forgets [eg](https://github.com/ocadotechnology/aimmo/pull/153/files#diff-d04a139bc2877a24168490f27b3091a2R83).